packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp

```diff
@@ -0,0 +1,378 @@
+            lno_t* out_array;
+            //out_degree_list is the offset array for indexing the out_array
+            lno_t* out_degree_list;
+
+            //member function to return the degree of a given vertex.
+            lno_t out_degree(lno_t vert){
+              return (out_degree_list[vert+1] - out_degree_list[vert]);
+            }
+            //member function to return the neighbors for a given vertex
+            lno_t* out_vertices(lno_t vert){
+              return (&out_array[out_degree_list[vert]]);
+            }
+        }; 
+        
+
+        //Enum that denotes
+	enum IcePropGrounding_Status {ICEPROPGS_FULL=2, ICEPROPGS_HALF=1, ICEPROPGS_NONE = 0};
+	// Struct representing a vertex label.
+	// We define our own "addition" for these labels.
+	// Later, we'll create a Tpetra::FEMultiVector of these labels.
+	template<typename lno_t, typename gno_t>
+	class IcePropVtxLabel {
+	public: 
+          //The local ID for the vertex represented by this label
+          lno_t id;
+          
+          //the global ID of a grounded vertex
+	  gno_t first_label;
+          
+          //the global ID of the vertex that sent the label
+	  gno_t first_sender;
+          
+          //this field indicates whether or not the first label is currently set
+          //(this is necessary for unsigned global types)
+          bool  first_used;
+          
+          //the global ID of a second grounded vertex
+	  gno_t second_label;
+
+          //the global ID of the vertex that sent the second identifier
+	  gno_t second_sender;
+
+          //this field indicates whether of not the second label is currently set
+          //(this is necessary for unsigned global types)
+          bool  second_used;
+
+          //this field may potentially be used to find biconnected components
+          int bcc_name;
+
+          //this flag indicates whether or not this vertex is a potential articulation point
+	  bool is_art;
+
+	  // Constructors
+	  IcePropVtxLabel(int idx_, int first_ = -1, int first_sender_ = -1, bool first_used_ = false, 
+                                    int second_ = -1, int second_sender_ = -1, bool second_used_ = false,
+                                    bool art_ = false, int bcc_name_ = -1) { 
+	    id = idx_;
+	    if(id == -1){
+		std::cout<<"A label's ID is -1\n";
+	    }
+	    first_label = first_;
+	    first_sender = first_sender_;
+            first_used = first_used_;
+	    second_label = second_;
+	    second_sender = second_sender_; 
+            second_used = second_used_;
+	    is_art = art_;
+            bcc_name = bcc_name_;
+	  }
+	  IcePropVtxLabel() {
+	    id = -1;
+	    first_label = -1;
+	    first_sender = -1;
+            first_used = false;
+	    second_label = -1;
+	    second_sender = -1;
+            second_used = false;
+	    is_art = false; 
+            bcc_name = -1;
+	  }
+
+	  IcePropVtxLabel(volatile const IcePropVtxLabel& other){
```

Comment:
Add comment about why all these variants of copy constructors and assignment overloads were needed.

Reply:
Done.